### PR TITLE
Zenduty

### PR DESCRIPTION
**Pull request by iNikitaGricenko**
### What changes are being made and why?

Changed body payload for zenduty integration api #94 (Removed bearer autzhorization #93)
---
### How the changes have been QAed?

Created account on Zenduty, and uses below example flow was sended alert
```yaml
id: zenduty
namespace: io.kestra.tests

listeners:
  - tasks:
      - id: zenduty
        type: io.kestra.plugin.notifications.zenduty.ZendutyExecution
        url: "https://www.zenduty.com/api/incidents/"
        message: "Kestra Zenduty alert"
        alertType: ERROR
        token: {{Zenduty_Token}}

tasks:
  - id: seq
    type: io.kestra.core.tasks.flows.Sequential
    tasks:
    - id: failed
      type: io.kestra.core.tasks.executions.Fail

```

---

### CODE REVIEW

File: src/main/java/io/kestra/plugin/notifications/zenduty/ZendutyTemplate.java [hunk 1]
```diff
@@ -38,40 +39,34 @@ public abstract class ZendutyTemplate extends ZendutyAlert {
     protected Map<String, Object> templateRenderMap;
 
     @Schema(
-        title = "Incident object's title"
+        title = "Incident title"
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/notifications/zenduty/ZendutyTemplate.java:42]:
```suggestion
        title = "Incident message"
```

File: src/main/java/io/kestra/plugin/notifications/zenduty/ZendutyTemplate.java [hunk 2]
```diff
@@ -38,40 +39,34 @@ public abstract class ZendutyTemplate extends ZendutyAlert {
...
-        title = "Escalation Policy object's unique_id"
+        title = "Incident type"
     )
     @PluginProperty(dynamic = true)
-    protected String escalationPolicy;
+    protected AlertType alertType;
 
     @Schema(
-        title = "SLA object's unique_id"
+        title = "Unique id for alert"
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/notifications/zenduty/ZendutyTemplate.java:60]:
```suggestion
        title = "Unique id for the alert"
```

File: src/main/java/io/kestra/plugin/notifications/zenduty/ZendutyTemplate.java [hunk 3]
```diff
@@ -38,40 +39,34 @@ public abstract class ZendutyTemplate extends ZendutyAlert {
...
     )
     @PluginProperty(dynamic = true)
-    protected String sla;
+    protected String entityId;
 
     @Schema(
-        title = "Priority object's unique_id"
+        title = "Array urls related to alert"
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/notifications/zenduty/ZendutyTemplate.java:66]:
```suggestion
        title = "List of URLs related to the alert"
```

This is not very descriptive, if you have more information about what those URL are please add it.

File: src/main/java/io/kestra/plugin/notifications/zenduty/ZendutyTemplate.java [hunk 4]
```diff
@@ -38,40 +39,34 @@ public abstract class ZendutyTemplate extends ZendutyAlert {
...
     )
     @PluginProperty(dynamic = true)
-    protected String assignedTo;
+    protected String summary;
 
     @Schema(
-        title = "Escalation Policy object's unique_id"
+        title = "Incident type"
     )
     @PluginProperty(dynamic = true)
```
Comment (loicmathieu) [inline src/main/java/io/kestra/plugin/notifications/zenduty/ZendutyTemplate.java:56]:
```suggestion
    @PluginProperty
```

Only strings can be rendered (also remove the runContext.render above)

File: src/main/resources/zenduty-template.peb [hunk 5]
```diff
@@ -1,22 +1,15 @@
...
-        "title": "{{title}}",
+    {% if message is defined %}
+        "message": "{{message}}",
+    {% endif %}
+    "alert_type": "{% if alertType is defined %}{{alertType}}{% elseif firstFailed == false %}resolved{% else %}error{% endif %}",
+    "entity_id": "{% if entityId is defined %}{{entityId}}{% else %}{{execution.id}}{% endif %}",
+    {% if urls is defined %}
+        "urls": [
+              {% for url in urls %}
+                      {{url | json}},
```
Comment (loicmathieu) [inline src/main/resources/zenduty-template.peb:10]:
```suggestion
                      {{url}},
```

I don't understand why you add the json filter here as URLs are string in the properties

**Review by loicmathieu — CHANGES_REQUESTED** [review]

**Comment by anna-geller** [comment]
After rebuilding the plugin 3 times and hard-refreshing my browser 10 times, I still see alertType being set as string, not Enum. Can you validate? 
image

File: src/main/java/io/kestra/plugin/notifications/zenduty/ZendutyExecution.java [hunk 6]
```diff
@@ -38,6 +38,7 @@
                     title: "Execution failed"
                     bearerAuth: xxx000yyy111
```
Comment (anna-geller) [inline src/main/java/io/kestra/plugin/notifications/zenduty/ZendutyExecution.java:39]:
```suggestion
```

no longer needed after adding the token

**Review by anna-geller — CHANGES_REQUESTED** [review]
The `ZendutyExecution` task somehow failed for me. Can you validate if you can reproduce it? You can ping me on Slack if you want to discuss or jump into a huddle 👍 thanks so much

```yaml
id: zenduty
namespace: dev

description: |
  The first issue is that alertType is still marked as a string for me without the typical autocompletion you'd get with Enum.

  Then, when using the events endpoint e.g. https://www.zenduty.com/api/events/fdbd2687-c319-4c3f-a924-fae6d83ce929/, the default payload configuration is wrong:
  Send Zenduty webhook: {"alert_type":"info","entity_id":"6mJ0bQdQNsecpVjWSoRDqG","summary":"Namespace dev, Flow ID hello_zen, Execution ID 6mJ0bQdQNsecpVjWSoRDqG, Execution Status SUCCESS. *<http://localhost:5173/ui/executions/dev/hello_zen/6mJ0bQdQNsecpVjWSoRDqG|[dev] hello_zen ➛ SUCCESS>*\n> Succeded after 00:00:02.228"}
  2023-10-23 22:41:07.820{"error": "Invalid payload"}

  When using the incident endpoint https://www.zenduty.com/api/incidents/, , the default flow from the PR body gives 500 error:

  Send Zenduty webhook: {"alert_type":"info","entity_id":"6Ag34ThLcMWEZto5ml2xF7","summary":"Namespace dev, Flow ID hello_zen, Execution ID 6Ag34ThLcMWEZto5ml2xF7, Execution Status SUCCESS. *<http://localhost:5173/ui/executions/dev/hello_zen/6Ag34ThLcMWEZto5ml2xF7|[dev] hello_zen ➛ SUCCESS>*\n> Succeded after 00:00:01.667"}
    Server Error (500)

tasks:
  - id: send_alert
    type: io.kestra.plugin.notifications.zenduty.ZendutyExecution
    url: "https://www.zenduty.com/api/incidents/" # https://www.zenduty.com/api/events/fdbd2687-c319-4c3f-a924-fae6d83ce929/
    # title: "Kestra flow failed"
    token: xxxxx
    alertType: INFO
    entityId: "{{ trigger.executionId }}"
    executionId: "{{ trigger.executionId }}"

triggers:
  - id: failed_prod_workflows
    type: io.kestra.core.models.triggers.types.Flow
    conditions:
      - type: io.kestra.core.models.conditions.types.ExecutionStatusCondition
        in:
          - SUCCESS
          - WARNING
      - type: io.kestra.core.models.conditions.types.ExecutionNamespaceCondition
        namespace: dev
        prefix: true
```

image

---

when using the events payload:
image

**Comment by iNikitaGricenko** [comment]
I did fix body a little bit (template), change documentation to links to new (correct) api and example code in documentation. Did make test with Zenduty account:
image

**Comment by iNikitaGricenko** [comment]
Seems like @anna-geller, you missed "message" field in payload when uses events payload

**Comment by anna-geller** [comment]
the message is an optional attribute?

image

and it should be optional, the goal of Execution tasks is to send default alert notifications

**Comment by anna-geller** [comment]
I think this is the last part to do @iNikitaGricenko to change a default message to something like:

`message: Kestra workflow execution {{ trigger.executionId }} of a flow {{ trigger.flowId }} in the namespace {{ trigger.namespace }} changed status to {{ trigger.state }}`

atm without setting `message`, it results in an invalid payload, even though the task property is marked as optional. so we need to set a default if the user didn't provide one. after that, we can merge this and I can make some small docs adjustments afterward.

**Comment by iNikitaGricenko** [comment]
> the message is an optional attribute?
> 
> image
> 
> and it should be optional, the goal of Execution tasks is to send default alert notifications

Message it is weird name of field from Zenduty. Actually it is title, so is required

**Review by anna-geller — APPROVED** [review]
Thanks so much!

**Pull request merged**